letsencrypt/acme/messages2.py

```diff
@@ -0,0 +1,276 @@
+       all challenges are presented without 'uri', 'status', or
+       'validated' fields.
+
+    """
+
+    __slots__ = ('chall',)
+    uri = jose.Field('uri')
+    status = jose.Field('status', decoder=Status.from_json)
+    validated = fields.RFC3339Field('validated', omitempty=True)
+
+    def to_json(self):
+        jobj = super(ChallengeBody, self).to_json()
+        jobj.update(self.chall.to_json())
+        return jobj
+
+    @classmethod
+    def fields_from_json(cls, jobj):
+        jobj_fields = super(ChallengeBody, cls).fields_from_json(jobj)
+        jobj_fields['chall'] = challenges.Challenge.from_json(jobj)
+        return jobj_fields
+
+
+class AuthorizationResource(Resource):
+    """Authorization resource.
+
+    :ivar body: `Authorization`
+    :ivar new_cert_uri: URI found in the 'next' Link header
+
+    """
+    __slots__ = ('body', 'uri', 'new_cert_uri')
+
+
+class Authorization(ResourceBody):
+    """Authorization resource body.
+
+    :ivar challenges: `list` of `Challenge`
+
+    """
+
+    identifier = jose.Field('identifier', decoder=Identifier.from_json)
+    challenges = jose.Field('challenges', omitempty=True)
+    combinations = jose.Field('combinations', omitempty=True)
+
+    # TODO: acme-spec #92, #98
+    key = Registration._fields['key']
+    contact = Registration._fields['contact']
+
+    status = jose.Field('status', omitempty=True, decoder=Status.from_json)
+    # TODO: 'expires' is allowed for Authorization Resources in
+    # general, but for Key Authorization '[t]he "expires" field MUST
+    # be absent'... then acme-spec gives example with 'expires'
+    # present... That's confusing!
+    expires = fields.RFC3339Field('expires', omitempty=True)
+
+    @challenges.decoder
+    def challenges(value):  # pylint: disable=missing-docstring,no-self-argument
+        return tuple(ChallengeBody.from_json(chall) for chall in value)
+
+    @property
+    def resolved_combinations(self):
+        """Combinations with challenges instead of indices."""
+        return tuple(tuple(self.challenges[idx] for idx in combo)
+                     for combo in self.combinations)
+
+
+class CertificateRequest(jose.JSONObjectWithFields):
+    """ACME new-cert request.
+
+    :ivar csr: `M2Crypto.X509.Request` wrapped in `.ComparableX509`
+
+    """
+    csr = jose.Field('csr', decoder=jose.decode_csr, encoder=jose.encode_csr)
+    authorizations = jose.Field('authorizations', decoder=tuple)
+
+
+class CertificateResource(Resource):
+    """Authorization resource.
+
+    :ivar body: `M2Crypto.X509.X509` wrapped in `.ComparableX509`
+    :ivar cert_chain_uri: URI found in the 'up' Link header
+    :ivar authzrs: `list` of `AuthorizationResource`.
+
+    """
+    __slots__ = ('body', 'uri', 'cert_chain_uri', 'authzrs')
+
+
+class Revocation(jose.JSONObjectWithFields):
+    """Revocation message.
+
+    :ivar revoke: Either a `datetime.datetime` or `NOW`.
```

Comment:
might be reasonable to say Revocation.NOW or change Revocation.NOW == 'NOW' just to avoid any confusion.